staging/src/k8s.io/apiserver/pkg/storage/cacher/btree_store.go

```diff
@@ -0,0 +1,189 @@
+	lock sync.RWMutex
+	tree *btree.BTree
+	n    int
+}
+
+func newBtreeStore(degree int) *btreeStore {
+	return &btreeStore{
+		tree: btree.New(degree),
+	}
+}
+
+func (t *btreeStore) Add(obj interface{}) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	return t.addOrUpdateLocked(obj, true)
+}
+
+func (t *btreeStore) Update(obj interface{}) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	return t.addOrUpdateLocked(obj, false)
+}
+
+func (t *btreeStore) Delete(obj interface{}) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	storeElem, ok := obj.(storeElement)
+	if !ok {
+		return fmt.Errorf("obj not a storeElement: %#v", obj)
+	}
+	item := t.tree.Delete(storeElem)
+	if item == nil {
+		return fmt.Errorf("obj does not exist")
+	}
+	t.n--
+
+	return nil
+}
+
+func (t *btreeStore) List() []interface{} {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	items := make([]interface{}, 0, t.n)
+	t.tree.Ascend(func(i btree.Item) bool {
+		items = append(items, i.(interface{}))
+		return true
+	})
+
+	return items
+}
+
+func (t *btreeStore) ListKeys() []string {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	items := make([]string, 0, t.n)
+	t.tree.Ascend(func(i btree.Item) bool {
+		items = append(items, i.(storeElement).Key)
+		return true
+	})
+
+	return items
+}
+
+func (t *btreeStore) Get(obj interface{}) (item interface{}, exists bool, err error) {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	storeElem, ok := obj.(storeElement)
+	if !ok {
+		return nil, false, fmt.Errorf("obj is not a storeElement")
+	}
+	item = t.tree.Get(storeElem)
+	if item == nil {
+		return nil, false, nil
+	}
+	return item, false, nil
+}
+
+func (t *btreeStore) GetByKey(key string) (item interface{}, exists bool, err error) {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	t.tree.Ascend(func(i btree.Item) bool {
+		if key == i.(storeElement).Key {
+			item = i
+			exists = true
+			return false
+		}
+		return true
+	})
+	return item, exists, nil
+}
+
+func (t *btreeStore) Replace(objs []interface{}, _ string) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	t.tree.Clear(false)
+	t.n = 0
+	for _, obj := range objs {
+		err := t.addOrUpdateLocked(obj, true)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (t *btreeStore) Resync() error {
+	// Nothing to do.
+	return nil
+}
+
+func (t *btreeStore) ListSince(rv uint64) []interface{} {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	items := []interface{}{}
+	t.tree.AscendGreaterOrEqual(storeElement{RV: rv}, func(i btree.Item) bool {
+		if rv < i.(storeElement).RV {
+			items = append(items, i.(interface{}))
+			return true
+		}
+		return false
+	})
+
+	return items
+}
+
+// addOrUpdateLocked performs the same operation but increments an
+// internal counter that keeps track of the number of elements that
+// the cache holds based on whether the operation is an add or update.
+// It is assumed to be called under a held lock.
+func (t *btreeStore) addOrUpdateLocked(obj interface{}, isAdd bool) error {
```

Comment:
Do we really need isAdd param?

My understand of ReplaceOrInsert is that it returns the previous value of the items.
So if it's an Add, it returns nil, otherwise it returns a real object.
So we can decide based on that.